Add RecordFunctionGuard

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36291 Make profiler thread local
* **#36215 Add RecordFunctionGuard**

Summary:
Make it possible to disable observers, e.g. to avoid
infinite recursion if an observer uses an operator

Test Plan:
USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py develop install
./build/bin/test_jit

Differential Revision: [D20912676](https://our.internmc.facebook.com/intern/diff/D20912676)

Commit:
Add RecordFunctionGuard

Summary:
Make it possible to disable observers, e.g. to avoid
infinite recursion if an observer uses an operator

Test Plan:
USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py develop install
./build/bin/test_jit

[ghstack-poisoned]

rohan-varma commented on torch/csrc/autograd/record_function.cpp:
```diff
@@ -218,7 +218,7 @@ void _runBeforeCallbacks(RecordFunction* rf, const std::string& funcName) {
 
 RecordFunction::RecordFunction(RecordScope scope) : scope_(scope) {
-  if (manager().hasCallbacks()) {
+  if (manager().hasCallbacks() && at::_tls_is_record_function_enabled()) {
```
By default, if there is no guard, this will always be true right?

ilia-cher replied:
correct

rohan-varma commented:
Is there an example where an infinite recursion may happen?

ilia-cher commented:
@rohan-varma yes, I believe it can happen e.g. in our NaN checker that works in observer - e.g. when you call torch::isnan it uses aten ops inside that ,in turn, call observers and so on; so we need to be careful not to call observers from observers code

Commit:
Update on "Add RecordFunctionGuard"


Summary:
Make it possible to disable observers, e.g. to avoid
infinite recursion if an observer uses an operator

Test Plan:
USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py develop install
./build/bin/test_jit

Differential Revision: [D20912676](https://our.internmc.facebook.com/intern/diff/D20912676)

[ghstack-poisoned]

Commit:
Update on "Add RecordFunctionGuard"


Summary:
Make it possible to disable observers, e.g. to avoid
infinite recursion if an observer uses an operator

Test Plan:
USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py develop install
./build/bin/test_jit

Differential Revision: [D20912676](https://our.internmc.facebook.com/intern/diff/D20912676)

[ghstack-poisoned]

Commit:
Update on "Add RecordFunctionGuard"


Summary:
Make it possible to disable observers, e.g. to avoid
infinite recursion if an observer uses an operator

Test Plan:
USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py develop install
./build/bin/test_jit

Differential Revision: [D20912676](https://our.internmc.facebook.com/intern/diff/D20912676)

[ghstack-poisoned]

Commit:
Update on "Add RecordFunctionGuard"


Summary:
Make it possible to disable observers, e.g. to avoid
infinite recursion if an observer uses an operator

Test Plan:
USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py develop install
./build/bin/test_jit

Differential Revision: [D20912676](https://our.internmc.facebook.com/intern/diff/D20912676)

[ghstack-poisoned]

Commit:
Update on "Add RecordFunctionGuard"


Summary:
Make it possible to disable observers, e.g. to avoid
infinite recursion if an observer uses an operator

Test Plan:
USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py develop install
./build/bin/test_jit

Differential Revision: [D20912676](https://our.internmc.facebook.com/intern/diff/D20912676)

[ghstack-poisoned]

Commit:
Update on "Add RecordFunctionGuard"


Summary:
Make it possible to disable observers, e.g. to avoid
infinite recursion if an observer uses an operator

Test Plan:
USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py develop install
./build/bin/test_jit

Differential Revision: [D20912676](https://our.internmc.facebook.com/intern/diff/D20912676)

[ghstack-poisoned]
